RecoEgamma/EgammaMCTools/plugins/PFClusterMatchedToPhotonsSelector.cc

```diff
@@ -0,0 +1,167 @@
+  trackingParticleToken_ = consumes<TrackingParticleCollection>(iConfig.getParameter<edm::InputTag>("trackingParticleTag"));
+  genParticleToken_ = consumes<reco::GenParticleCollection>(iConfig.getParameter<edm::InputTag>("genParticleTag"));
+
+  matchMaxDR_      = iConfig.getParameter<double>("maxDR");
+  volumeZ_EB_      = iConfig.getParameter<double>("volumeZ_EB");
+  volumeRadius_EB_ = iConfig.getParameter<double>("volumeRadius_EB");
+  volumeZ_EE_      = iConfig.getParameter<double>("volumeZ_EE");
+
+  produces<reco::PFClusterCollection>();
+  produces<reco::PFCluster::EEtoPSAssociation>();
+  produces<edm::ValueMap<reco::GenParticleRef> >();
+}
+
+
+void PFClusterMatchedToPhotonsSelector::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("pfClustersTag", edm::InputTag("particleFlowClusterECAL"));
+  desc.add<edm::InputTag>("trackingParticleTag", edm::InputTag("mix", "MergedTrackTruth"));
+  desc.add<edm::InputTag>("genParticleTag", edm::InputTag("genParticles"));
+  desc.add<double>("maxDR", 0.3);
+  desc.add<double>("volumeZ_EB", 304.5);
+  desc.add<double>("volumeRadius_EB", 123.8);
+  desc.add<double>("volumeZ_EE", 317.0);
+  descriptions.add("pfClusterMatchedToPhotonsSelector", desc);
+
+}
+
+void PFClusterMatchedToPhotonsSelector::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+
+  edm::Handle<reco::PFClusterCollection> particleFlowClusterECALHandle_;
+  edm::Handle<reco::PFCluster::EEtoPSAssociation> associationHandle_;
+  edm::Handle<TrackingParticleCollection> trackingParticleHandle_;
+  edm::Handle<reco::GenParticleCollection> genParticleHandle_;
+  iEvent.getByToken(particleFlowClusterECALToken_, particleFlowClusterECALHandle_);
+  iEvent.getByToken(trackingParticleToken_, trackingParticleHandle_);  
+  iEvent.getByToken(genParticleToken_, genParticleHandle_);  
+  iEvent.getByToken(associationToken_, associationHandle_);  
+
+  std::unique_ptr<reco::PFClusterCollection>                      out   = std::make_unique<reco::PFClusterCollection>();
+  std::unique_ptr<reco::PFCluster::EEtoPSAssociation> association_out   = std::make_unique<reco::PFCluster::EEtoPSAssociation>();
+  std::unique_ptr<edm::ValueMap<reco::GenParticleRef> > genmatching_out = std::make_unique<edm::ValueMap<reco::GenParticleRef> >();
+
+  std::vector<reco::GenParticleRef> genmatching;
+
+  size_t iP(0);
+  size_t iN(0);  
+  for (auto&& pfCluster : *particleFlowClusterECALHandle_) {
+
+    bool isMatched = false;
+    reco::GenParticleRef::key_type matchedKey;
+
+    for (auto&& trackingParticle : *trackingParticleHandle_) {
+      if (trackingParticle.pdgId() != 22) continue;
+      if (trackingParticle.status() != 1) continue;
+      matchedKey = trackingParticle.genParticles().at(0).key();
+      if (reco::deltaR(trackingParticle, pfCluster.position()) > matchMaxDR_) continue; 
```

Comment:
@perrotta Ok, I fixed. I also added the DEDR cut that I had the variable in the code but never used. In EGM, we do these studies in particle gun samples, so the matching is quite simple (there is only object in the event), but it is good to have an energy consistency as well to improve the matching against pileup.